cmd/zoekt-sourcegraph-indexserver/fetcher.go

```diff
@@ -0,0 +1,167 @@
+}
+
+type branchDeltas map[string]delta
+
+// delta contains all the paths that have been added, changed, and removed
+// along with the file contents
+//
+// TODO@ggilmore: should the file contents  be io.ReadClosers, and should delta offer a "close"
+// method that closes all the file contents (thinking about the local file case)
+type delta struct {
+	// Added is a list of all the paths that have been added,
+	// along with their file contents.
+	Added map[string]io.Reader
+
+	// Changed is a list of all the paths that have been changed,
+	// along with their file contents.
+	Changed map[string]io.Reader
+
+	// Removed is a list of all the paths that were removed.
+	Removed []string
+}
+
+// NOTE: We can mimic the existing behavior (albeit with extra work on Zoekt's end)
+// by creating a fresh git checkout every time, feeding it to local fetcher,
+// and copying all the files into a new repository for zoekt-git-index to process. This
+// doesn't seem that useful, but we can do this to be able to merge a PR earlier.
+//
+// The delta will appear as if all files were deleted and then re-added on every commit.
+//
+// We can have another service whose job is just to manage a permanent local checkout (deal with corruption,
+// garbage collection, evicting epos, etc.) but the implementation of localFetcher doesn't
+// care about where this service comes from.
+
+// localFetcher is a deltaFetcher that operates against a local git checkout
+type localFetcher struct {
+	// gitDir is the local file path to the .git folder for the repository
+	gitDir string
+
+	// remoteURL is the URL for the git remote that commits are fetched from
+	remoteURL string
+
+	// customRunCommand, if provided, is used to execute all constructed
+	// git commands instead of just cmd.Run().
+	customRunCommand func(*exec.Cmd) error
+}
+
+// Fetch returns the paths that have changed on the latest
+// commit for each branch, along with their file contents.
+func (lf *localFetcher) Fetch(ctx context.Context, branches ...zoekt.RepositoryBranch) (branchDeltas, error) {
+	runCmd := func(cmd *exec.Cmd) error {
+		return cmd.Run()
+	}
+	if lf.customRunCommand == nil {
+		runCmd = lf.customRunCommand
+	}
+
+	gitCmd := func(args ...string) *exec.Cmd {
+		args = append([]string{"git", "-C", lf.gitDir, "-c"}, args...)
+		c := exec.CommandContext(ctx, "git", args...)
+		c.Stdin = &bytes.Buffer{}
+		return c
+	}
+
+	// Create branches in local checkout if they don't exist
+	for _, b := range branches {
+		// Does the local branch exist?
+		ref := fmt.Sprintf("refs/heads/%s", b.Name)
+		cmd := gitCmd("show-ref", "--verify", ref)
+
+		var errB bytes.Buffer
+		cmd.Stderr = &errB
+		if err := runCmd(cmd); err != nil {
+			if !strings.Contains(errB.String(), "not a valid ref") {
+				return nil, fmt.Errorf("while checking to see if branch %q exists in local checkout: %w", b.Name, err)
+			}
+
+			// create the missing branch
```

Comment:
mutating your checkout and adding a branch seems like it could lead to bad behaviour if the checkout is used by others. Can we not mutate the inputs so the branch disappears? This is what we do when generating the list of zoekt.RepositoryBranches. In fact if the input list is coming from sourcegraph, you should error out if checking out fails since there is then bad state / a rare race that will disappear on next invocation.